Initial proposal (#1)
w3c/trace-context-mqtt: spec/21-MQTT3_FORMAT.md

> # MQTT 3 recommendation

Author: Comment from @zteo-google

In general I think it's not a good idea to modify a user payload to carry the trace context. This can cause breakages in parsers and would be considered a disruptive feature. We should treat the user payload as a black box that should not be tampered with.

My recommendation is that traces be carried on a completely separate MQTT topic, preferably one that is considered 'system reserved' in MQTT (for eg. topics $SYS/traces or something similar). For packet correlation purposes, either the packet ID (QoS 1 and 2), or a digest of the original message can be used.

Author: @zteo-google this section is mentioned as non-normative. Meaning it's just a recommendation for how it can be done if one decided to do it

Reviewer: ```suggestion
# MQTT v3 recommendation
```

Before MQTT 5 there were no extensibility metadata defined for the messages. Thus
this section is only a set of recommendations on how trace context can be
implemented, not the specification.

## JSON payload

> It is common that MQTT message payload represented as JSON string. In this case

Reviewer: Grammar clean

```suggestion
It is common that MQTT message payloads are represented as a JSON string. In this case
```

it is recommended to set `traceparent` and `tracestate` fields as a fields on a
outermost JSON object.

For example, if original payload looks like following:

``` json
{
    "_type": "cmd",
    "action": "setConfiguration",
    "configuration": {
        "_type":"configuration",
        "settingA": "value A",
        "settingB": 42
        }
}
```

Payload with trace context configured will look like following:

``` json
{
    "_type": "cmd",
    "action": "setConfiguration",
    "configuration": {
        "_type":"configuration",
        "settingA": "value A",
        "settingB": 42
        },
    "traceparent": "00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
    "tracestate": "congo=BleGNlZWRzIHRohbCBwbGVhc3VyZS4"
}
```

> ## Handling e2e encryption

Author: From @zteo-google

End-to-end encryption is generally handled at the transport protocol layer so message payloads are typically not "additionally encrypted". I understand the issue here, and it is similar to the binary payload issue in the paragraph below. Again, my recommendation is not to touch payloads and instead to publish to a separate MQTT topic.

Reviewer: ```suggestion
## Handling end-to-end encryption
```

TBD: Should there be a note that for e2e encryption one may decide to send trace
context un-encrypted while the rest of the message can be encrypted?

## Binary payload

> For binary payload - [binary protocol](..\extension-binary.html) may be used to

Author: the right link is https://github.com/w3c/trace-context-binary/pull/2
